PwaUriHandler/explainer.md

```diff
@@ -0,0 +1,250 @@
+
+#### Handling multiple registrations
+
+The association file is able to contain association objects for multiple PWA handlers. Multiple PWAs are able to request to handle the same URI. In cases where there are multiple registered handlers available, the browser or OS should present options to the user and allow the user to decide which handler to use.
+
+### Operating System Changes
+
+To provide OS level registration of URI handlers, browsers' PWA implementations need to integrate more deeply with the OS application platform. This is to allow the OS to recognize PWAs as apps that can be launched and can therefore serve as URI handlers. There may be different approaches to accomplishing this, and they might differ on different OSes. OS changes may be necessary to enable this integration. In OS or browser versions where this integration cannot be implemented, the behavior should default to URI handling by the browser.
+
+## Security Considerations
+
+URI handlers, if implemented improperly, can be used to hijack traffic for websites. This is why the pwa to site association mechanism is an important part of the proposed scheme. Associations must be validated when PWAs are installed, and they should be validated again periodically to evaluate any new changes in the association file and web app manifest.
+
+In the case where the browser registers URI handling with the OS on behalf of PWAs, the OS must trust the browser to validate the PWA-to-site association or implement the validation itself. If the OS delegates the validation to browsers, it must be clear which browsers are compliant with the validation requirements.
+
+If an associated site is overtaken by a malicious actor, it is possible for users to be exposed to malicious content through the PWA handling those URIs. To mitigate this risk, the browser may want to suppress the PWA launch or get user confirmation using a security mechanism which detects risky URIs.
+
+Conforming browsers may want to limit the maximum allowed number of `uri_handlers` entries to N and the numbers of `paths` and `excludePaths` each to M. This will limit the amount of work the manifest parser does and further limit the risk of URI hijacking.
+
+URI handler registrations should only be performed for installed PWAs as users expect installed applications to be more deeply integrated with the OS. Furthermore, conforming browsers should not register PWAs as URI handlers for any URI without an explicit user confirmation.
+
+## Privacy Considerations
+
+The introduction of URI handling features carries a fingerprinting risk: a third party could build up a profile on a user if they can determine which PWAs that the user has installed on their device.
```

Comment:
Is the third party in this paragraph another website/PWA (as opposed to a native app)? Can you add more detail as to how this would be accomplished?

Reply:
The third party can be another website or a native app. What we're trying to express in the following two paragraphs is that third party websites cannot trigger this feature, and native applications on the user's system already have access to this information anyway (with the user's consent).

Any suggestions for how I can improve the language here? We went back and forth on this section many times.

Reply:
Ah, I see. I think we could:

* Cut this first paragraph
* Update the second one to say something like "As websites are currently unable to launch URI handlers through normal in-browser navigation, we do not anticipate third-party websites being able to leverage this API to detect which PWAs the user has installed."

It's possible there could be a clever attack I haven't thought of yet. Anything that springs to mind currently is fairly convoluted, would depend on interesting cookie jar partitioning, and the site already having a good idea who you are anyway. :)

Reply:
Edited.